wpilibc/src/main/native/cpp/simulation/DutyCycleEncoderSim.cpp

```diff
@@ -16,12 +16,49 @@ DutyCycleEncoderSim::DutyCycleEncoderSim(int channel) {
   frc::sim::SimDeviceSim deviceSim{"DutyCycle:DutyCycleEncoder", channel};
   m_simPosition = deviceSim.GetDouble("position");
   m_simDistancePerRotation = deviceSim.GetDouble("distance_per_rot");
+  m_simAbsolutePosition = deviceSim.GetDouble("absPosition");
+  m_simIsConnected = deviceSim.GetBoolean("connected");
+}
+
+double DutyCycleEncoderSim::Get() {
+  return m_simPosition.Get();
 }
 
 void DutyCycleEncoderSim::Set(units::turn_t turns) {
   m_simPosition.Set(turns.value());
+  m_simAbsolutePosition.Set(std::fmod(turns.value(), 1));
+}
+
+double DutyCycleEncoderSim::GetDistance() {
+  return m_simPosition.Get() * m_simDistancePerRotation.Get();
 }
 
 void DutyCycleEncoderSim::SetDistance(double distance) {
-  m_simPosition.Set(distance / m_simDistancePerRotation.Get());
+  double pos = distance / m_simDistancePerRotation.Get();
+  m_simPosition.Set(pos);
+  m_simAbsolutePosition.Set(std::fmod(pos, 1));
```

Comment:
Is this the correct behavior, or does it need to be offset from the current m_simAbsolutePosition value?

Reply:
Depends on whether or not it's relative to last reset. Actually, the position and absolute position methods change depending on whether or not we make them relative to last reset.

Reply:
Taking a step back... we need to consider three things: 1) how does the real hardware operate in response to the API, 2) what's the "real world" view of the sensor, and 3) what's an intuitive simulation-facing interface/API?  For DutyCycleEncoder, the API gives both an absolute position and a relative position (by tracking rollovers of the absolute position).  I don't think it makes sense to *only* expose absolute position at the sim level and track rollovers on it like the real hardware does, so exposing absolute and relative separately makes sense.  The question is whether it's actually desirable from a sim perspective for the absolute and relative values to automatically interact within the sim layer or whether they should just be completely independent and leave it up to the user simulation bits to set them both appropriately.

Reply:
I think it's better if we leave absolute and relative values up to the user. If a user sets absolute position, they'd need to set relative position, so it's not really easier on the user if position automatically sets absolute position.

Reply:
Makes sense to me.  In which case this line should be removed.